File: packages/a2/src/agent/file-system.ts
/**
 * @license
 * Copyright 2025 Google LLC
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  DataPart,
  DeepReadonly,
  FileDataPart,
  InlineDataCapabilityPart,
  Outcome,
  StoredDataCapabilityPart,
} from "@breadboard-ai/types";
import { err, ok } from "@breadboard-ai/utils";
import mime from "mime";

export { AgentFileSystem };

const KNOWN_TYPES = ["audio", "video", "image", "text"];

export type FileDescriptor = {
  type: "text" | "storedData" | "inlineData" | "fileData";
  mimeType: string;
  data: string;
  title?: string;
  resourceKey?: string;
};

export type AddFilesToProjectResult = {
  existing: string[];
  added: string[];
  total: number;
  error?: string;
};

class AgentFileSystem {
  #fileCount = 0;

  #projects: Map<string, Set<string>> = new Map();

  #files: Map<string, FileDescriptor> = new Map();

  write(data: string, mimeType: string): string {
    const name = this.create(mimeType);
    this.#files.set(name, { data, mimeType, type: "text" });
    return name;
  }

  #getFile(path: string): Outcome<DataPart> {
    const file = this.#files.get(path);
    if (!file) {
      return err(`file "${path}" not found`);
    }
    switch (file.type) {
      case "fileData":
        return {
          fileData: {
            fileUri: file.data,
            mimeType: file.mimeType,
            resourceKey: file.resourceKey,
          },
        } satisfies FileDataPart;
      case "inlineData":
        return {
          inlineData: {
            data: file.data,
            mimeType: file.mimeType,
            title: file.title,
          },
        } satisfies InlineDataCapabilityPart;
      case "storedData":
        return {
          storedData: {
            handle: file.data,
            mimeType: file.mimeType,
            resourceKey: file.resourceKey,
          },
        } satisfies StoredDataCapabilityPart;
      default:
      case "text":
        return {
          text: file.data,
        };
    }
  }

  #getProjectFiles(path: string): Outcome<DataPart[]> {
    const project = this.#projects.get(path);
    if (!project) {
      return err(`Project "${project}" not found`);
    }
    const errors: string[] = [];
    const files = [...project].map((path) => {
      const file = this.#getFile(path);
      if (!ok(file)) {
        errors.push(file.$error);
      }
      return file;
    });
    if (errors.length > 0) {
      return err(errors.join(","));
    }
    return files as DataPart[];
  }

  get(path: string): Outcome<DataPart[]> {
    // Do a path fix-up just in case: sometimes, Gemini decides to use
    // "vfs/file" instead of "/vfs/file".
    if (path.startsWith("vfs/")) {
      path = `/${path}`;
    }
    if (path.startsWith("/vfs/projects")) {
      return this.#getProjectFiles(path);
    }
    const file = this.#getFile(path);
    if (!ok(file)) return file;
    return [file];
  }

  createProject(name: string): string {
    return `/vfs/projects/${name}`;
  }

  addFilesToProject(
    projectPath: string,
    files: string[]
  ): AddFilesToProjectResult {
    let project = this.#projects.get(projectPath);
    if (!project) {
      project = new Set();
      this.#projects.set(projectPath, project);
    }
    const existing = [...project];
    files.forEach((file) => project.add(file));
    return {
      total: project.size,
      existing,
      added: files,
    };
  }

  listProjectContents(projectPath: string): string[] {
    const project = this.#projects.get(projectPath);
    return [...(project || [])];
  }

  add(part: DataPart): Outcome<string> {
    if ("text" in part) {
      const mimeType = "text/markdown";
      const name = this.create(mimeType);
      this.#files.set(name, { type: "text", mimeType, data: part.text });
      return name;
    } else if ("inlineData" in part) {
      const { mimeType, data, title } = part.inlineData;
      const name = this.create(mimeType);
      this.#files.set(name, { type: "inlineData", mimeType, data, title });
      return name;
    } else if ("storedData" in part) {
      const { mimeType, handle: data, resourceKey } = part.storedData;
      const name = this.create(mimeType);
      this.#files.set(name, {
        type: "storedData",
        mimeType,
        data,
        resourceKey,
      });
      return name;
    } else if ("fileData" in part) {
      const { mimeType, fileUri: data, resourceKey } = part.fileData;
      const name = this.create(mimeType);
      this.#files.set(name, { type: "fileData", mimeType, data, resourceKey });
      return name;
    }
    return err(`Unsupported part: ${JSON.stringify(part)}`);
  }

  get files(): ReadonlyMap<string, DeepReadonly<FileDescriptor>> {
    return this.#files;
  }

  create(mimeType: string) {
    const name = this.#getName(mimeType);
    const ext = mime.getExtension(mimeType);
    return `/vfs/${name}${++this.#fileCount}.${ext}`;
  }

  #getName(mimeType: string) {
    const first = mimeType.split("/").at(0) || "";
    if (KNOWN_TYPES.includes(first)) return first;
    return "file";
  }
}

File: packages/bees/bees/protocols/filesystem.py
# Copyright 2026 Google LLC
# SPDX-License-Identifier: Apache-2.0

"""Bees-native FileSystem protocol and supporting types.

These mirror the shapes in ``opal_backend.file_system_protocol`` so that
``DiskFileSystem`` can import from here instead. Python's structural
subtyping means opal_backend's concrete types satisfy these definitions
without modification.

See ``spec/filesystem.md`` for design rationale.
"""

from __future__ import annotations

import mimetypes
from dataclasses import dataclass
from typing import Any, Callable, Protocol, runtime_checkable

# ---------------------------------------------------------------------------
# MIME type registration
# ---------------------------------------------------------------------------

# Ensure common MIME types are registered — mirrors the side effect in
# opal_backend.file_system_protocol.
mimetypes.add_type("text/markdown", ".md")
mimetypes.add_type("text/csv", ".csv")

# ---------------------------------------------------------------------------
# Constants
# ---------------------------------------------------------------------------

KNOWN_TYPES = ["audio", "video", "image", "text"]
DEFAULT_EXTENSION = "txt"
DEFAULT_MIME_TYPE = "text/plain"

# ---------------------------------------------------------------------------
# Type aliases
# ---------------------------------------------------------------------------

SystemFileGetter = Callable[[], str | dict[str, str]]
"""A callable that returns file content or an error dict."""

# ---------------------------------------------------------------------------
# Data types
# ---------------------------------------------------------------------------


@dataclass
class FileDescriptor:
    """Describes a file stored in an agent file system."""

    data: str
    mime_type: str
    type: str  # "text", "inlineData", "fileData", "storedData"
    title: str | None = None
    resource_key: str | None = None


@dataclass
class FileSystemSnapshot:
    """Serializable snapshot of file system state.

    Contains everything needed to reconstruct a file system from
    persisted state. For disk-backed implementations this is a
    point-in-time capture of what's on disk.

    Transient state (system file getters) is re-attached by the caller
    after reconstruction.
    """

    files: dict[str, FileDescriptor]
    routes: dict[str, str]
    file_count: int


# ---------------------------------------------------------------------------
# Utility functions
# ---------------------------------------------------------------------------


def file_descriptor_to_part(file: FileDescriptor) -> dict[str, Any]:
    """Convert a FileDescriptor to a Gemini data part dict.

    Handles ``fileData``, ``inlineData``, ``storedData``, and text types.
    """
    if file.type == "fileData":
        return {
            "fileData": {
                "fileUri": file.data,
                "mimeType": file.mime_type,
                **(
                    {"resourceKey": file.resource_key}
                    if file.resource_key
                    else {}
                ),
            }
        }
    if file.type == "inlineData":
        return {
            "inlineData": {
                "data": file.data,
                "mimeType": file.mime_type,
                **({"title": file.title} if file.title else {}),
            }
        }
    if file.type == "storedData":
        return {
            "storedData": {
                "handle": file.data,
                "mimeType": file.mime_type,
                **(
                    {"resourceKey": file.resource_key}
                    if file.resource_key
                    else {}
                ),
            }
        }
    # Default: text
    return {"text": file.data}


# ---------------------------------------------------------------------------
# Protocol
# ---------------------------------------------------------------------------


@runtime_checkable
class FileSystem(Protocol):
    """Protocol for agent file systems — in-memory or disk-backed.

    Consumers should depend on this protocol, not on a concrete class.
    """

    def add_system_file(self, path: str, getter: SystemFileGetter) -> None:
        """Register a virtual system file backed by a getter function."""
        ...

    def overwrite(self, name: str, data: str) -> str:
        """Write (or overwrite) a named text file. Returns the path."""
        ...

    def write(self, name: str, data: str) -> str:
        """Write a named file. Returns the path."""
        ...

    def append(self, path: str, data: str) -> dict[str, str] | None:
        """Append data to an existing text file, or create it.

        Returns ``None`` on success, or an error dict on failure.
        """
        ...

    async def read_text(self, path: str) -> str | dict[str, str]:
        """Read the text content of a file.

        Returns the text string, or an error dict if not found / not text.
        """
        ...

    async def get(self, path: str) -> list[dict[str, Any]] | dict[str, str]:
        """Get the data parts for a file path.

        Returns a list of Gemini data parts, or an error dict.
        """
        ...

    async def get_many(
        self, paths: list[str],
    ) -> list[dict[str, Any]] | dict[str, str]:
        """Get data parts for multiple file paths.

        Returns all parts, or an error dict with joined errors.
        """
        ...

    async def list_files(self) -> str:
        """List all files as newline-separated paths."""
        ...

    def get_file_url(self, maybe_path: str) -> str | None:
        """Get the URL for a file path, if it has a displayable URL.

        Returns ``None`` for text files or unknown paths.
        """
        ...

    def add_part(
        self, part: dict[str, Any], file_name: str | None = None,
    ) -> str | dict[str, str]:
        """Add a data part to the file system. Returns the path or error dict."""
        ...

    def add_route(self, original_route: str) -> str:
        """Register a route and return its pidgin name."""
        ...

    def get_original_route(self, route_name: str) -> str | dict[str, str]:
        """Look up the original route for a pidgin route name."""
        ...

    @property
    def files(self) -> dict[str, FileDescriptor]:
        """Read-only access to all stored files."""
        ...

    @property
    def snapshot(self) -> FileSystemSnapshot:
        """Capture serializable state."""
        ...
